site/source/docs/porting/Rust.rst

```diff
@@ -0,0 +1,51 @@
+.. Rust:
+
+=================
+Rust + Emscripten
+=================
+
+The
+`Rust programming language <https://www.rust-lang.org/>`_
+has direct support for compiling to WebAssembly.
+In addition to that, Rust can be used with Emscripten, which allows Rust to
+benefit from Emscripten's API support. For example, using Emscripten with Rust
+can allow you to port a Rust codebase using :ref:`OpenGL <OpenGL-support>` or
+:ref:`Pthreads <Pthreads-support>`, as Emscripten has library support code for
+those APIs which handles the large differences between them and the closest Web
+APIs.
+
+============
+Installation
+============
+
+The Rust toolchain will look for ``emcc`` and the other Emscripten components
+in the path. You should therefore first
+:ref:`download, install, and add Emscripten to your path <sdk-download-and-install>`.
+
+It is crucial to install the proper versions of the tools. Specifically, Rust
+uses a particular LLVM version, and Emscripten must use the same one, so that
+object files and so forth are all compatible.
+
+As of June 2021, stable Rust is ``1.52.1``, which uses LLVM 12, which was
+branched off of the ``llvmorg-12-init`` tag. That LLVM change rolled into
+Emscripten in
+`56b877afe7d1b651d6b9a2ba5d5df074876172ff <https://chromium.googlesource.com/emscripten-releases/+/56b877afe7d1b651d6b9a2ba5d5df074876172ff>`_,
+which means that the compatible version of Emscripten can be installed as follows:
+
+    .. code-block:: bash
+
+       emsdk install 56b877afe7d1b651d6b9a2ba5d5df074876172ff
+       emsdk activate 56b877afe7d1b651d6b9a2ba5d5df074876172ff
```

Comment:
It would be worth finding and documenting the first tagged Emscripten release after this point, IMO, rather than suggesting that users install an untagged release. A priori, I would expect that the risk of LLVM incompatibilities would be outweighed by the downsides of documenting an untagged git commit, namely that the commit is impossible to remember and raises questions about stability for users who do not know about our release process.

Reply:
This is the hash of a roll, so it has passed the roller's full test suite. For that reason I don't think the risk is significantly higher than a release.

It worries me to allow several further rolls in between (which is what going to the next release would cause). It's not just LLVM incompatibilities, it's binaryen as well, where LLVM and binaryen overlap (say, on the binary format of a new feature).

I think the minimal possible distance in terms of commits, that is still fully tested, is safest.

Reply:
As far as I understand, we can add arbitrary tags to hashes in emsdk pretty easily. Could we add a tag for this version like `rust-1.52.1` or `llvm-12` to make it easier to use & remember?

Reply:
Looks like it just needs a tag here? https://github.com/emscripten-core/emsdk/blob/main/emscripten-releases-tags.txt

Reply:
Good idea! Yes, it is as simple as that.

+1 for adding both of those tags there, as both could be useful for people wanting to use emscripten with a provided LLVM.

We should also test them. Testing the LLVM one can be done with a PR in emscripten (that replaces the tot build with LLVM 12). For Rust, I'm not sure how best to do it, but one option might be to create an `emscripten-core/rust` repo and add some amount of testing there. Or perhaps the repo could be more generically named, as we may want multiple languages there eventually (e.g. Zig).

Reply:
> Yes, it is as simple as that.

(narrator voice) It wasn't as simple as that 😅 https://github.com/emscripten-core/emsdk/pull/836